### ISSUE TITLE
Build system does not natively support universal binaries for macOS

### ISSUE THREAD

**Issue by crazydef** [randombit/botan#3517]
Hi,

I'm trying to build a universal (static) library on macOS (Ventura on an M2 Pro) but I'm getting the above error.

It's entirely possible what I'm attempting (`--cc-abi-flags="-force_cpusubtype_ALL -arch x86_64 -arch arm64"`) isn't possible, but, well, it probably should be.  :)

Anyway, this is the error I'm getting:
```
error: unknown target CPU 'armv8+crypto'
note: valid target CPU values are: nocona, core2, penryn, bonnell, atom, silvermont, slm, goldmont, goldmont-plus, tremont, nehalem, corei7, westmere, sandybridge, corei7-avx, ivybridge, core-avx-i, haswell, core-avx2, broadwell, skylake, skylake-avx512, skx, cascadelake, cooperlake, cannonlake, icelake-client, rocketlake, icelake-server, tigerlake, sapphirerapids, alderlake, knl, knm, k8, athlon64, athlon-fx, opteron, k8-sse3, athlon64-sse3, opteron-sse3, amdfam10, barcelona, btver1, btver2, bdver1, bdver2, bdver3, bdver4, znver1, znver2, znver3, x86-64, x86-64-v2, x86-64-v3, x86-64-v4
make: *** [build/obj/lib/block_aes_armv8.o] Error 1
```

**Comment by reneme** [comment]
Probably some internal architecture selection in `./configure.py` is interfering with the compiler flags you selected manually via `--cc-abi-flags`. I'll look into that on my M2.

**Comment by mouse07410** [comment]
First, I noticed that Universal often fails on Apple Silicon Macs, totally unrelated to Botan.

Second, I question the wisdom of building universal - what, you have two different processors on your computer?! Or you can't compile for x86 arch on Intel Mac? Or you can't afford time to compile twice and copy the resulting libraries to where they belong? And you have extra disk space that you're eager to consume?

I could understand why Apple does it - perhaps it's simpler in the long run to have one OS binary package for both platforms. I rather doubt your distribution needs are on the Apple's scale.

**Comment by reneme** [comment]
A quick workaround: if you run `configure.py` with `--disable-armv8crypto` you essentially disable the usage of ARMv8 crypto extensions altogether. That way, your build should not fail anymore, because the culprit compiler flag (`-march=armv8+crypto`) won't be used anymore.

As a side note: `configure.py` will disable x86 crypto extensions for you automagically. The configure script is not aware of any universal builds and will skip x86 extensions because your host system is ARM.

**Comment by crazydef** [comment]
First, that's not been my experience.  Everything I'm building right now (both for work and personal projects) is built for both architectures, and I've never had any problems.

Secondly, it's how every other library and project I have here is built.  (And not having to have special cases/double builds just for macOS makes it easier to manage - I only need to juggle seven platforms instead of eight.)

Thirdly, for our in-house teams (for work projects) providing a single executable means they don't have to care what they install.  Or rather, we don't have to figure out what their machine is and install the right binaries.  Ditto for publishing software for download by external users (my own projects).  Providing a single package for macOS will simplify my life, and my user's lives.

`As a side note: configure.py will disable x86 crypto extensions for you automagically. The configure script is not aware of any universal builds and will skip x86 extensions because your host system is ARM.`

Does that mean it's not possible for me to build a non-universal, optimised Intel build on my Arm machine even if I wanted to?

**Comment by reneme** [comment]
> Does that mean it's not possible for me to build a non-universal, optimised Intel build on my Arm machine even if I wanted to?

Personally I didn't try that so far. But as a typical cross-compile this should almost certainly be possible without any problems.

However the `configure.py` is not aware of universal builds out of the box. As a result, it will take decisions based on your build host.

For a more sophisticated workaround you could try to build the library as x86 and ARM (each with crypto extensions enabled) independently. Then combine the resulting binaries with `lipo`. I didn't try that yet, though.

**Comment by randombit** [comment]
I think the solution in the past has been to use `lipo` to combine two binaries. That at least is how it is documented and IIRC several people have reported success doing it that way.

> Secondly, it's how every other library and project I have here is built.

I think the trick here is that most libraries don't use any hardware specific interfaces. If you're content with that, what you're doing currently can probably be accomplished using `--cpu=generic`.

**Comment by reneme** [comment]
I had a quick glance into our build scripts to see if we could natively support the "dual build followed by the final `lipo` invocation. Unfortunately that would be quite a project to pull off. Additionally, it would probably lack the flexibility for people to use effectively as well.

**Comment by crazydef** [comment]
Yeah, I saw in the build instructions that it probably wouldn't work.  But I'm always optimistic about these things, and was quietly hoping the documentation was just out of date.  :)

Thanks for the suggestions, but in the end I decided to just bump my schedule and update my system to use Apple's Network framework.  (It was something I was going to do eventually anyway - Botan was only ever intended to be a temporary solution on Apple platforms that I could use to get the Mac port of my engine (personal project) up and running quickly.)
